[feat] a11y-no-redundant-roles

jamesb93 opened the pull request:
This solves one task on the list in #820. My first PR so probably a lot going on wrong here but I would be very appreciative for feedback and guidance on contributing.

### Before submitting the PR, please make sure you do the following
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] Prefix your PR title with `[feat]`, `[fix]`, `[chore]`, or `[docs]`.
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
-  [x] Run the tests with `npm test` and lint the project with `npm run lint`

bluwy commented on src/compiler/compile/nodes/Element.ts:
```diff
@@ -351,6 +351,53 @@ export default class Element extends Node {
+				// no-redundant-roles
+				if (this.name === value) {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				}
+				if (this.name === 'nav' && value === 'navigation') {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				}
+
+				if (this.name === 'a' && value === 'link') {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				}
+
+				if (this.name === 'fieldset' && value === 'group') {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				}
+
+				if (this.name === 'ul' && value === 'list') {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				}
```
Should we use `else if` here?

jamesb93 replied:
Definitely.

bluwy commented on src/compiler/compile/nodes/Element.ts:
```diff
@@ -351,6 +351,53 @@ export default class Element extends Node {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				}
+
+				if (this.name === 'main' && value === 'main') {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				}
```
I think this is already covered above with `this.name === value`

jamesb93 replied:
Ah yes, thanks!

bluwy commented on src/compiler/compile/nodes/Element.ts:
```diff
@@ -351,6 +351,53 @@ export default class Element extends Node {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				}
+
+				let { parent } = this;
+				let is_parent_section_or_article = false;
+
+				while (parent) {
+					const parent_name = (parent as Element).name;
+					if ( parent_name === 'section' || parent_name === 'article') {
+						is_parent_section_or_article = true;
+						break;
+					}
+					if (parent.type === 'Element') {
+						break;
+					}
+					parent = parent.parent;
+				}
```
I think we can extract this out as a function.

bluwy commented on src/compiler/compile/nodes/Element.ts:
```diff
@@ -351,6 +351,53 @@ export default class Element extends Node {
+					parent = parent.parent;
+				}
+
+				if (this.name === 'header' && value === 'banner' && !is_parent_section_or_article) {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				}
+
+				if (this.name === 'footer' && value === 'contentinfo' && !is_parent_section_or_article) {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				}
```
We could use `else if` here too, and maybe have `!is_parent_section_or_article` as the first bigger if check, before checking for the names and values.

bluwy commented on src/compiler/compile/compiler_warnings.ts:
```diff
@@ -76,6 +76,10 @@ export default {
 		code: 'a11y-unknown-role',
 		message: `A11y: Unknown role '${role}'` + (suggestion ? ` (did you mean '${suggestion}'?)` : '')
 	}),
+	a11y_no_redundant_roles: (role: string | boolean , suggestion?: string) => ({
+		code: 'a11y-no-redundant-roles',
+		message: `A11y: Redundant role '${role}'` + (suggestion ? ` (did you mean '${suggestion})?'` : '')
+	}),
```
I doesn't seem like the `suggestion` param is used. I think we can removed that.

jamesb93 replied:
Yep. I Initially imagined maybe it would suggest an alternative but the number of different scenarios a reundant role can appear in might be quite large and require different advice.

jamesb93 commented:
Thanks for your feedback @bluwy. Youre suggestions and enhancements have been implemented in db606c7, 91ec3fc, 53b8919, 17906c2, 649e462.

bluwy commented on src/compiler/compile/nodes/Element.ts:
```diff
@@ -351,6 +368,28 @@ export default class Element extends Node {
 				}
+
+				// no-redundant-roles
+				if (this.name === value) {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				} else if (this.name === 'nav' && value === 'navigation') {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				} else if (this.name === 'a' && value === 'link') {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				} else if (this.name === 'fieldset' && value === 'group') {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				} else if (this.name === 'ul' && value === 'list') {
+					component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+				}
```
I just noticed that they are repeating `component.warn(...` lines. I think we can merge them into one, and have a big if condition for it, which might be more readable.

jamesb93 replied:
It's not great to read I agree, maybe the conditions could be stored somewhere and a single check is out in place against those conditions? I can whip that up shortly.

bluwy replied:
That could be a solution too, I'd personally keep it as an if condition so the logic is clear upfront. But I'm indifferent of either solutions, you can implement whichever you prefer 🙂

jamesb93 replied:
You're right :) No need to seriously over-engineer things when readability could be preserved. `git commit`s incoming.

jamesb93 replied:
Pushed in bb975e7

bluwy commented on src/compiler/compile/nodes/Element.ts:
```diff
@@ -351,6 +368,28 @@ export default class Element extends Node {
+
+				const is_parent_section_or_article = is_parent(this.parent, ['section', 'article']);
+				if (!is_parent_section_or_article) {
+					if (this.name === 'header' && value === 'banner') {
+						component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+					} else if (this.name === 'footer' && value === 'contentinfo') {
+						component.warn(attribute, compiler_warnings.a11y_no_redundant_roles(value));
+					}
```
Probably same here as well for consistency

jamesb93 replied:
This is now addressed in bb975e7

dummdidumm commented:
I just realized there exists a PR already for this: #5361 . Though I'm not sure which implementation should be preferred.

bluwy commented:
Crap now I feel bad for choosing either one since both has its good parts and effort put into it. #5361 does have more checks. On the other hand, this PR also handles parent section/article checks. If I have to choose, I think we should continue this, and borrow over the extra checks from #5361 to here. We can then have @mhatvan as the co-author in the commit description.

jamesb93 commented:
> Crap now I feel bad for choosing either one since both has its good parts and effort put into it. #5361 does have more checks. On the other hand, this PR also handles parent section/article checks. If I have to choose, I think we should continue this, and borrow over the extra checks from #5361 to here. We can then have @mhatvan as the co-author in the commit description.

I'd be more than happy to update with work already done in https://github.com/sveltejs/svelte/pull/5361. I think the joining of the two would make for a robust set of checks.

jamesb93 commented:
I've updated this PR with the conditions that are checked in PR #5361 @bluwy + @dummdidumm. I've also updated the logic so that there isn't a massive `if` statement with numerous nested `||`. Instead there is now a `Map()` that contains the many cases that can arise with anything else being caught by `this.name === value`.

bluwy reviewed (commented):
Thanks for further improving the PR! I've left some comments to help with performance, but after that this looks good to me.

jamesb93 commented on src/compiler/compile/nodes/Element.ts:
```diff
@@ -351,6 +368,62 @@ export default class Element extends Node {
+				const nested_implicit_semantics = new Map([
+					['header', 'banner'],
+					['footer', 'contentinfo']
+				]);
```
Addressed in 2b05e54.

jamesb93 commented on src/compiler/compile/nodes/Element.ts:
```diff
@@ -351,6 +368,62 @@ export default class Element extends Node {
+					['thead', 'rowgroup'],
+					['tr', 'row'],
+					['ul', 'list']
+				]);
```
Addressed in 2b05e54.

bluwy commented on src/compiler/compile/nodes/Element.ts:
```diff
@@ -70,6 +70,47 @@ const a11y_labelable = new Set([
+
+const a11y_implicit_semantics = new Map([
+	['a', 'link'],
+	['article', 'article'],
```
I think this can be removed.

jamesb93 replied:
No problem - many requested changes is good for me learning and the contribution process :)

I've made the change in 6ab5c7fa6fdc0f75a5c82f89d801f80c00728284

bluwy reviewed (approved):
Thanks for keeping up with the many requested changes 😅 LGTM with one tiny nitpick.

mhatvan commented:
Well, I tried to get attention to my PR without success for a while.  Anyway, I am fine with having this PR merged with stuff taken from mine ;)
The most important part is to get more a11y checks into svelte!

jamesb93 commented:
> Well, I tried to get attention to my PR without success for a while. Anyway, I am fine with having this PR merged with stuff taken from mine ;) The most important part is to get more a11y checks into svelte!

It is definitely important to acknowledge that a significant portion of the code was championed by you @mhatvan. I learned from it and it looks like the checks should be pretty robust as a result. Thank you :)

mhatvan commented:
Thanks man, I appreciate it!

All that matters is more and improved a11y checks ;)

jamesb93 commented:
It seems like one specific test (unrelated perhaps to this PR) is failing on macOS-latest and node 12. Is there any guidance on making it so this PR is fully test compliant?

dummdidumm commented:
Don't worry about that, according to CI they just timed out.

Conduitry commented:
Released in 3.45.0 - thank you!

jamesb93 commented:
> Released in 3.45.0 - thank you!

Woo! Thanks everyone for your guidance and hard work. Much appreciated and I learned a lot.
